File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/throttle/ThrottleData.java
/*
 *  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package org.wso2.micro.gateway.core.throttle;

import java.util.concurrent.atomic.AtomicLong;

/**
 * Class to hold throttle counter data against a specific key
 */
public class ThrottleData {

    /**
     * Enum to hold throttle types supported by gateway.
     */
    public enum ThrottleType {
        APP, SUBSCRIPTION, RESOURCE
    }

    private long windowStartTime = 0;
    private long unitTime;
    private AtomicLong count = new AtomicLong();
    private long remainingQuota;
    private boolean stopOnQuota = true;
    private boolean throttled = false;
    private String throttleKey;
    private ThrottleType throttleType;

    public long getWindowStartTime() {
        return windowStartTime;
    }

    public void setWindowStartTime(long windowStartTime) {
        this.windowStartTime = windowStartTime;
    }

    public long getUnitTime() {
        return unitTime;
    }

    public void setUnitTime(long unitTime) {
        this.unitTime = unitTime;
    }

    public AtomicLong getCount() {
        return count;
    }

    public void setCount(AtomicLong count) {
        this.count = count;
    }

    public long getRemainingQuota() {
        return remainingQuota;
    }

    public void setRemainingQuota(long remainingQuota) {
        this.remainingQuota = remainingQuota;
    }

    public boolean isStopOnQuota() {
        return stopOnQuota;
    }

    public void setStopOnQuota(boolean stopOnQuota) {
        this.stopOnQuota = stopOnQuota;
    }

    public boolean isThrottled() {
        return throttled;
    }

    public void setThrottled(boolean throttled) {
        this.throttled = throttled;
    }

    public ThrottleType getThrottleType() {
        return throttleType;
    }

    public void setThrottleType(ThrottleType throttleType) {
        this.throttleType = throttleType;
    }

    public void setThrottleKey(String throttleKey) {
        this.throttleKey = throttleKey;
    }

    public boolean cleanThrottleData(long timeStamp) {
        if (this.windowStartTime + this.unitTime < timeStamp) {
            switch (getThrottleType()) {
                case APP: {
                    ThrottleCounter.removeFromApplicationCounterMap(this.throttleKey);
                    break;
                }
                case RESOURCE: {
                    ThrottleCounter.removeFromResourceCounterMap(this.throttleKey);
                    break;
                }
                case SUBSCRIPTION: {
                    ThrottleCounter.removeFromSubscriptionCounterMap(this.throttleKey);
                    break;
                }
            }
            return true;
        }
        return false;
    }
}

File: filter-chain/filters/throttleData.go
package filters

import (
	_ "github.com/cactus/go-statsd-client/statsd"
)

type ThrottleType int

const (
	APP ThrottleType = iota
	SUBSCRIPTION
	RESOURCE
	API
)

func (t ThrottleType) String() string {
	return [...]string{"APP", "SUBSCRIPTION", "RESOURCE", "API"}[t]
}

type ThrottleData struct {
	windowStartTime int64
	unitTime        int64
	count           int64
	remainingQuota  int64
	stopOnQuota     bool
	throttled       bool
	throttleKey     string
	throttleType    ThrottleType
}

func (t ThrottleData) getWindowStartTime() int64 {
	return t.windowStartTime
}

func (t *ThrottleData) setWindowStartTime(windowStartTime int64) {
	t.windowStartTime = windowStartTime
}

func (t ThrottleData) getUnitTime() int64 {
	return t.unitTime
}

func (t *ThrottleData) setUnitTime(unitTime int64) {
	t.unitTime = unitTime
}

func (t ThrottleData) getCount() int64 {
	return t.count
}

func (t *ThrottleData) setCount(count int64) {
	t.count = count
}

func (t ThrottleData) getRemainingQuota() int64 {
	return t.remainingQuota
}

func (t *ThrottleData) setRemainingQuota(remainingQuota int64) {
	t.remainingQuota = remainingQuota
}

func (t ThrottleData) isStopOnQuota() bool {
	return t.stopOnQuota
}

func (t *ThrottleData) setStopOnQuota(stopOnQuota bool) {
	t.stopOnQuota = stopOnQuota
}

func (t ThrottleData) isThrottled() bool {
	return t.throttled
}

func (t *ThrottleData) setThrottled(throttled bool) {
	t.throttled = throttled
}

func (t ThrottleData) getThrottleType() ThrottleType {
	return t.throttleType
}

func (t *ThrottleData) setThrottleType(throttleType ThrottleType) {
	t.throttleType = throttleType
}

func (t *ThrottleData) setThrottleKey(throttleKey string) {
	t.throttleKey = throttleKey
}

func cleanThrottleData(t ThrottleData, timeStamp int64) bool {
	if t.windowStartTime+t.unitTime < timeStamp {
		switch throttleType := t.getThrottleType(); throttleType {
		case APP:
			{
				removeFromApplicationCounterMap(t.throttleKey)
				break
			}
		case API:
			{
				removeFromApiCounterMap(t.throttleKey)
				break
			}
		case RESOURCE:
			{
				removeFromResourceCounterMap(t.throttleKey)
				break
			}
		case SUBSCRIPTION:
			{
				removeFromSubscriptionCounterMap(t.throttleKey)
				break
			}
		}
		return true
	}
	return false
}
